plans/sep6.md

```diff
@@ -0,0 +1,213 @@
+}
+
+class PendingDeposit extends PendingTransfer {
+  deposit: (args: DepositRequest) => Promise<Request>;
+}
+interface FeeArgs {
+  assetCode: string;
+  amount: string;
+  operation: "withdraw" | "deposit";
+  type: "string";
+}
+
+interface WithdrawAssetInfo {
+  assetCode: string;
+  fee: Fee;
+  minAmount: number;
+  maxAmount: number;
+  types: {
+    name: string;
+    fields: Field[];
+  }[];
+}
+
+interface WithdrawInfo {
+  [assetCode: string]: WithdrawAssetInfo;
+}
+
+interface DepositAssetInfo {
+  assetCode: string;
+  fee: Fee;
+  minAmount: number;
+  maxAmount: number;
+  fields: Field[];
+}
+
+interface DepositInfo {
+  [assetCode: string]: DepositAssetInfo;
+}
+
+interface WithdrawRequest {
+  type: string;
+  assetCode: string;
+  dest: string;
+  destExtra?: string;
+  account?: string;
+  memo: Memo;
+}
+
+interface DepositRequest {
+  assetCode: string;
+  account: string;
+  memo?: Memo;
+  emailAddress?: string;
+  type?: string;
+}
+
+type responseTypes =
+  | "ok"
+  | "non_interactive_customer_info_needed"
+  | "interactive_customer_info_needed"
+  | "customer_info_status"
+  | "error";
+
+interface TransferResponse {
+  type: responseTypes;
+}
+
+interface WithdrawOk extends TransferResponse {
+  status: "ok";
+  sendTo: string;
+  needsMemo: boolean;
+  memo: Memo;
+  eta?: number;
+  minAmount?: number;
+  maxAmount?: number;
+  fee: Fee;
+  extraInfo?;
+}
+
+interface DepositOk extends TransferResponse {
+  how: string;
+  eta?: number;
+  minAmount?: number;
+  maxAmount?: number;
+  feeFixed?: number;
+  feePercent?: number;
+  extraInfo?: {
+    message: string;
+  };
+}
+
+interface NonInteractiveKycNeeded extends TransferResponse {
+  type: "non_interactive_customer_info_needed";
+  fields: string[]; // This can be restricted to the list of strings in SEP-9
+  // https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0009.md#kyc--aml-fields
+}
+
+interface InteractiveKycNeeded extends TransferResponse {
+  type: "interactive_customer_info_needed";
+  url: string;
+  interactiveDeposit?: boolean;
+}
+
+interface KycStatus extends TransferResponse {
+  type: "customer_info_status";
+  more_info_url?: string;
+  eta?: number;
+}
+
+interface CustomerInfoStatus extends KycStatus {
+  status: "pending" | "denied";
+}
+
+interface KycPromptStatus extends KycStatus {
+  status: "success" | "pending" | "denied";
+}
+
+interface Field {
+  name: string;
+  description: string;
+  optional?: boolean;
+  choices?: [string];
+}
+
+interface Fee {
+  type: "none" | "simple" | "complex";
+}
+
+interface SimpleFee extends Fee {
+  type: "simple";
+  fixed: number;
+  percent: number;
+}
+
+interface Memo {
+  type: "text" | "id" | "hash";
+  value: string;
+}
+```
+
+### Usage
+
+```js
+import { startWithdrawal, startDeposit, RESPONSE_TYPES } from "wallet-sdk";
+
+// Retrieves /info and returns a class used to complete the withdrawal/deposit
+const withdrawal = await startWithdrawal(transferServerUrl);
```

Comment:
I think the name `startWithdrawal` doesn't really adequately explain what the function is returning, and `withdrawal` doesn't correctly describe what that object is. Really, this `withdrawal` object is a withdrawal-handling service, so I think this should be a `new` constructor (debatable), and it should be named similarly to DataProvider (I feel more strongly about this):

```js
const withdrawalProvider = new WithdrawalProvider(transferServerUrl);
```

There's also the consideration that "Withdrawal" could be confused with "Withdraw", and given the choice between two similar things, it might be better to go with the shorter of the two.

Reply:
Updated!